scripts/clean_osm_data.py

```diff
@@ -728,6 +728,29 @@ def filter_lines_by_geometry(df_all_lines):
     return df_all_lines
 
 
+def drop_external_shapes(df_region, df, only_boundary=False):
+    """
+    Function to drop external shapes from GeoDataFrames.
+
+    Parameters
+    ----------
+    df_region : GeoDataFrame
+        GeoDataFrame with the shape of the considered region
+    df : GeoDataFrame
+        GeoDataFrame to be cleaned
+    only_boundary : bool
+        If True, only the boundary of df is considered
+        If False, the whole geometry of df is considered
+    """
+    geom_obj = df.geometry.boundary if only_boundary else df.geometry
+    contain_all = gpd.sjoin(df_region, geom_obj.to_frame(), predicate="contains")
```

Comment:
@davide-f Great code, as usually :) 

I wonder, if ` drop_external_shapes()` being used to make the code cleaner or should it improve performance as well?

Another question: is it possible to keep `"within"` as the predicate? There is some evidence (e.g. [here](https://gis.stackexchange.com/a/175230/118363) or [here](https://github.com/geopandas/dask-geopandas/issues/114)) that `within` has very good performance.

Reply:
The fact is that the old code was doing two services in one go: 1) setting the country name by shape and 2) dropping overpassing objects
The use of the sjoin nearest only covers 1, but not 2.
Therefore there was the need to add that functionality in an efficient way, and I preferred a new function to make the code clean.

Actually contains is the converse of within: a.contains(b) === b.within(a) in shapely (by definition).
I expect sjoin to behave alike: sjoin(a,b, contains) === sjoin(b,a, within).
If I have time I can test.

I tested various combinations with shapely and alike and this was the best, but I don't think I tested that!
Thanks for the advice! :D I'll keep that in consideration!

Reply:
> The fact is that the old code was doing two services in one go: 1) setting the country name by shape and 2) dropping overpassing objects The use of the sjoin nearest only covers 1, but not 2. Therefore there was the need to add that functionality in an efficient way, and I preferred a new function to make the code clean.
> 
> Actually contains is the converse of within: a.contains(b) === b.within(a) in shapely (by definition). I expect sjoin to behave alike: sjoin(a,b, contains) === sjoin(b,a, within). If I have time I can test.
> 
> I tested various combinations with shapely and alike and this was the best, but I don't think I tested that! Thanks for the advice! :D I'll keep that in consideration!

Agree that now the code looks cleaner. So, functionality split between two function does the job perfectly :)

Regarding `within` vs `contains`, you are absolutely right that both `sjoin` should work exactly the same. Looking into sjoin() [source code](https://github.com/geopandas/geopandas/blob/91d18898d770254f460adf2260da238acb501553/geopandas/tools/sjoin.py#L203-L214), it seems that `within` is simply replaced by `contains`, as internal implementation of `contains` is [faster](https://github.com/geopandas/geopandas/pull/1421) for some reasons. But that is the detail, and from practical perspective there is no difference. So, I think no additional tests are needed ;)

Reply:
Ok :)